contracts/token/ERC20WithPermit.sol

```diff
@@ -2,6 +2,224 @@
+        _burn(msg.sender, amount);
+    }
+
+    function burnFrom(address account, uint256 amount) external override {
+        uint256 currentAllowance = allowance[account][msg.sender];
+        require(currentAllowance >= amount, "Burn amount exceeds allowance");
+        _approve(account, msg.sender, currentAllowance - amount);
+        _burn(account, amount);
+    }
+
+    function approveAndCall(
+        address spender,
+        uint256 value,
+        bytes memory extraData
+    ) external override returns (bool) {
+        if (approve(spender, value)) {
+            IReceiveApproval(spender).receiveApproval(
+                msg.sender,
+                value,
+                address(this),
+                extraData
+            );
+            return true;
+        }
+        return false;
+    }
+
+    function approve(address spender, uint256 amount)
+        public
+        override
+        returns (bool)
+    {
+        _approve(msg.sender, spender, amount);
+        return true;
+    }
+
+    /* solhint-disable-next-line func-name-mixedcase */
+    function DOMAIN_SEPARATOR() public view override returns (bytes32) {
+        if (chainId() == cachedChainId) {
+            return cachedDomainSeparator;
+        } else {
+            return buildDomainSeparator();
+        }
+    }
+
+    function _burn(address account, uint256 amount) internal {
+        uint256 currentBalance = balanceOf[account];
+        require(currentBalance >= amount, "Burn amount exceeds balance");
+        balanceOf[account] = currentBalance - amount;
+        totalSupply -= amount;
+        emit Transfer(account, address(0), amount);
+    }
+
+    function _transfer(
+        address sender,
+        address recipient,
+        uint256 amount
+    ) private {
+        require(sender != address(0), "Transfer from the zero address");
+        require(recipient != address(0), "Transfer to the zero address");
+        uint256 senderBalance = balanceOf[sender];
+        require(senderBalance >= amount, "Transfer amount exceeds balance");
+        balanceOf[sender] = senderBalance - amount;
+        balanceOf[recipient] += amount;
+        emit Transfer(sender, recipient, amount);
+    }
+
+    function _approve(
+        address owner,
+        address spender,
+        uint256 amount
+    ) private {
+        require(owner != address(0), "Approve from the zero address");
+        require(spender != address(0), "Approve to the zero address");
+        allowance[owner][spender] = amount;
+        emit Approval(owner, spender, amount);
+    }
+
+    function buildDomainSeparator() private view returns (bytes32) {
+        return
+            keccak256(
+                abi.encode(
+                    keccak256(
+                        "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+                    ),
+                    keccak256(bytes(name)),
+                    keccak256(bytes("1")),
```

Comment:
Maybe we want to pull the version to a constant defined on top of this contract or to be configurable on contract deployment?

Reply:
This value is strictly related to EIP712 domain we are creating in this function. It is the current major version of the signing domain and the domain we are constructing is the domain in version 1. This version is not something we want to configure when deploying.